asmcomp/loongarch64/emit.mlp

```diff
@@ -0,0 +1,773 @@
+  | Imod  -> "mod.d"
+  | _ -> Misc.fatal_error "Emit.Intop"
+
+let name_for_intop_imm = function
+  | Iadd -> "addi.d"
+  | Iand -> "andi"
+  | Ior  -> "ori"
+  | Ixor -> "xori"
+  | Ilsl -> "slli.d"
+  | Ilsr -> "srli.d"
+  | Iasr -> "srai.d"
+  | _ -> Misc.fatal_error "Emit.Intop_imm"
+
+let name_for_floatop1 = function
+  | Inegf -> "fneg.d"
+  | Iabsf -> "fabs.d"
+  | Ispecific Isqrtf -> "fsqrt.d"
+  | _ -> Misc.fatal_error "Emit.Iopf1"
+
+let name_for_floatop2 = function
+  | Iaddf -> "fadd.d"
+  | Isubf -> "fsub.d"
+  | Imulf -> "fmul.d"
+  | Idivf -> "fdiv.d"
+  | _ -> Misc.fatal_error "Emit.Iopf2"
+
+let name_for_specific = function
+  | Imultaddf false -> "fmadd.d"
+  | Imultaddf true  -> "fnmadd.d"
+  | Imultsubf false -> "fmsub.d"
+  | Imultsubf true  -> "fnmsub.d"
+  | _ -> Misc.fatal_error "Emit.Iopf3"
+
+(* Output the assembly code for an instruction *)
+
+let emit_instr env i =
+  emit_debug_info i.dbg;
+  match i.desc with
+    Lend -> ()
+  | Lprologue ->
+      assert (env.f.fun_prologue_required);
+      let n = frame_size env in
+      emit_stack_adjustment (-n);
+      if env.f.fun_contains_calls then begin
+        store_ra n;
+        cfi_offset ~reg:1 (* ra *) ~offset:(-8)
+      end;
+  | Lop(Imove | Ispill | Ireload) ->
+      let src = i.arg.(0) and dst = i.res.(0) in
+      if src.loc <> dst.loc then begin
+        match (src, dst) with
+        | {loc = Reg _; typ = (Val | Int | Addr)}, {loc = Reg _} ->
+            `	move      {emit_reg dst}, {emit_reg src}\n`
+        | {loc = Reg _; typ = Float}, {loc = Reg _; typ = Float} ->
+            `	fmov.d   {emit_reg dst}, {emit_reg src}\n`
+        | {loc = Reg _; typ = (Val | Int | Addr)}, {loc = Stack s} ->
+            let (base, ofs) = slot_offset env s (register_class dst) in
+            emit_store src ofs base
+        | {loc = Reg _; typ = Float}, {loc = Stack s} ->
+            let (base, ofs) = slot_offset env s (register_class dst) in
+            emit_float_store src ofs base
+        | {loc = Stack s; typ = (Val | Int | Addr)}, {loc = Reg _} ->
+            let (base, ofs) = slot_offset env s (register_class src) in
+            emit_load dst ofs base
+        | {loc = Stack s; typ = Float}, {loc = Reg _} ->
+            let (base, ofs) = slot_offset env s (register_class src) in
+            emit_float_load dst ofs base
+        | {loc = Reg _; typ = Float}, {loc = Reg _; typ = (Val | Int | Addr)}
+        | {loc = Stack _}, {loc = Stack _}
+        | {loc = Unknown}, _ | _, {loc = Unknown} ->
+            Misc.fatal_error "Emit: Imove"
+      end
+  | Lop(Iconst_int n) ->
+      `	li.d	{emit_reg i.res.(0)}, {emit_nativeint n}\n`
+  | Lop(Iconst_float f) ->
+      let lbl = new_label() in
+      env.float_literals <- {fl=f; lbl} :: env.float_literals;
+      `la.local {emit_reg reg_tmp}, {emit_label lbl} \n`;
+      `	fld.d	{emit_reg i.res.(0)}, {emit_reg reg_tmp}, 0\n`
+  | Lop(Iconst_symbol s) ->     (* FIXME la.global assert error in binutils*)
```

Comment:
Can you give more details about that binutils error? Does it still apply to recent binutils versions, such as 2.41?

Reply:
- assert error info

```
make -j6
/usr/bin/ld: BFD (GNU Binutils) 2.40.0 断言失败 elfnn-loongarch.c：3144
/usr/bin/ld: BFD (GNU Binutils) 2.40.0 断言失败 elfnn-loongarch.c：3144
/usr/bin/ld: BFD (GNU Binutils) 2.40.0 断言失败 elfnn-loongarch.c：3144
/usr/bin/ld: BFD (GNU Binutils) 2.40.0 断言失败 elfnn-loongarch.c：3144
/usr/bin/ld: BFD (GNU Binutils) 2.40.0 断言失败 elfnn-loongarch.c：3144
/usr/bin/ld: BFD (GNU Binutils) 2.40.0 断言失败 elfnn-loongarch.c：3144
/usr/bin/ld: BFD (GNU Binutils) 2.40.0 断言失败 elfnn-loongarch.c：3144
/usr/bin/ld: BFD (GNU Binutils) 2.40.0 断言失败 elfnn-loongarch.c：3144
/usr/bin/ld: BFD (GNU Binutils) 2.40.0 断言失败 elfnn-loongarch.c：3144
collect2: 错误：ld 返回 1
File "caml_startup", line 1:
Error: Error during linking (exit code 1)
```

- source code
wget https://ftp.gnu.org/gnu/binutils/binutils-2.40.tar.bz2

./bfd/elfnn-loongarch.c
```
3139     case R_LARCH_GOT_PC_HI20:
3140     case R_LARCH_GOT_HI20:
3141       /* Calc got offset.  */
3142         {
3143           unresolved_reloc = false;
3144           BFD_ASSERT (rel->r_addend == 0);
3145 
3146           bfd_vma got_off = 0;
3147           if (h != NULL)
3148         {
3149           /* GOT ref or ifunc.  */
3150           BFD_ASSERT (h->got.offset != MINUS_ONE
3151                   || h->type == STT_GNU_IFUNC);
3152 
3153           got_off = h->got.offset  & (~(bfd_vma)1);
3154           /* Hidden symbol not has got entry,
3155            * only got.plt entry so it is (plt - got).  */
3156           if (h->got.offset == MINUS_ONE && h->type == STT_GNU_IFUNC)
3157             {
3158               bfd_vma idx;
3159               if (htab->elf.splt != NULL)
3160             {
3161               idx = (h->plt.offset - PLT_HEADER_SIZE)
3162                 / PLT_ENTRY_SIZE;
3163               got_off = sec_addr (htab->elf.sgotplt)
3164                 + GOTPLT_HEADER_SIZE
3165                 + (idx * GOT_ENTRY_SIZE)
3166                 - sec_addr (htab->elf.sgot);
3167             }
3168               else
3169             {
3170               idx = h->plt.offset / PLT_ENTRY_SIZE;
3171               got_off = sec_addr (htab->elf.sgotplt)
3172                 + (idx * GOT_ENTRY_SIZE)
3173                 - sec_addr (htab->elf.sgot);
3174             }

```